climada_petals/entity/exposures/osm_dataloader.py

```diff
@@ -0,0 +1,346 @@
+        """
+
+        nodes_taken = []
+        data_geom = []
+        data_id = []
+        data_tags = []
+
+        for way in result.ways:
+            if way.id in ways_avail:
+                node_lat_lons = []
+                for node in way.nodes:
+                    nodes_taken.append(node.id)
+                    node_lat_lons.append((float(node.lat), float(node.lon)))
+                data_geom.append(shapely.geometry.LineString(node_lat_lons))
+                data_id.append(way.id)
+                data_tags.append(way.tags)
+
+            if closed_lines_are_polys:
+                data_geom = [shapely.geometry.Polygon(way) if way.is_closed
+                             else way for way in data_geom]
+
+        gdf_ways = gpd.GeoDataFrame(
+            data={'osm_id': data_id,'geometry': data_geom, 'tags':data_tags})
+
+        return nodes_taken, gdf_ways
+
+    def _assemble_from_nodes(self, result, nodes_avail):
+        """
+        pick out those nodes and ways from result instance that belong to
+        ways. Assemble ways into gdfs. Keep track of which nodes
+        and ways have been "used up" already
+
+        Parameters
+        ---------
+        result : overpy.Overpass result object
+        nodes_avail : list
+            node-ids that have not yet been used for relation and way
+            construction and are hence available for node constructions
+
+        Returns
+        -------
+        gdf_nodes : gpd.GeoDataFrame
+            gdf with nodes that were assembled from result object
+        """
+        data_geom = []
+        data_id = []
+        data_tags = []
+
+        for node in result.nodes:
+            if node.id in nodes_avail:
+                data_geom.append(shapely.geometry.Point(node.lat, node.lon))
+                data_id.append(node.id)
+                data_tags.append(node.tags)
+
+        gdf_nodes = gpd.GeoDataFrame(
+            data={'osm_id': data_id,'geometry': data_geom, 'tags':data_tags})
+
+        return gdf_nodes
+
+    def _update_availability(self, full_set, to_remove):
+        """
+        update id availabilities from whole result set after using up some for
+        geometry construction
+        """
+        return [item for item in full_set if item not in to_remove]
+
+    def _assemble_results(self, result, closed_lines_are_polys=True):
+        """
+        assemble an overpass result object with results, ways, nodes etc. from the
+        format-specific structure into one geodataframe
+
+        Parameters
+        ---------
+        result : overpy.Overpass result object
+        closed_lines_are_polys : bool
+                whether closed lines are polygons. Default is True
+
+        Returns
+        ------
+        gdf_results : gpd.GeoDataFrame
+            Result-gdf from the overpass query.
+        """
+        gdf_results = gdf_ways = gdf_nodes = gpd.GeoDataFrame(
+            columns=['osm_id','geometry','tags'])
```

Comment:
Similar to above. CRS is missing, geometry is not set explicitly. In general, I would say it does not help to initialize an empty GeoDataFrame for this reason. Simply take the ones returned by the other methods and concatenate them, instead of appending to the empty one.